### PR TITLE
Additional improvements to docs and makefile.

### PR DESCRIPTION
**Pull request by frankban**

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui-charm/114/
Test PASSed.

File: Makefile [hunk 1]
```diff
@@ -19,6 +19,9 @@ VENV = tests/.venv
 SYSDEPS = build-essential bzr charm-tools firefox libapt-pkg-dev \
 	libpython-dev python-virtualenv rsync xvfb
 
+CHARMCMD := charm
+CHARMURL ?= cs:~juju-gui-charmers/juju-gui
```
Comment (urosj) [inline Makefile]: Why non-promulgated space?

File: QA.md [hunk 2]
```diff
@@ -40,7 +40,7 @@ soon.
   make deploy
   ```
 
-* Deployed via the charm using juju v2.0 / 1.26alpha
+* Deployed via the charm using Juju v2.0 / 1.26alpha
```
Comment (urosj) [inline QA.md]: Drop 1.26alpja, just Juju 2.0.

File: Makefile [hunk 3]
```diff
@@ -94,4 +111,8 @@ help:
 	@echo '  the charm will be deployed in the default Juju environment.'
 	@echo '  If SERIES is not passed, "trusty" is used. Possible values are'
 	@echo -e '  "precise" and "trusty".\n'
+	@echo 'make publish-development - Push and publish the charm to the'
+	@echo -e '  development channel.\n'
+	@echo 'make publish-stable - Push and publish the charm to the'
+	@echo -e '  stable channel.\n'
```
Comment (urosj) [inline Makefile]: I'd rather remove publishing to stable through make script. This should always be manual.

File: QA.md [hunk 4]
```diff
@@ -72,11 +78,40 @@ All of the above tests need to be run on these providers:
...
-  - Do not test with juju 1.24
-  - juju 1.25 should allow model creation, pending GUI support.
-  - juju 2.0 will not allow model creation in the near term.
+  - Do not test with Juju 1.24
+  - Juju 1.25 should allow model creation, pending GUI support.
+  - Juju 2.0 will not allow model creation in the near term.
+
+## Automated functional testing ##
+
+The uitest project contains integration tests for many Juju services, including
```
Comment (urosj) [inline QA.md]: Uitests are a private repo.

**Comment by urosj** [comment]
Some suggestions there.

File: Makefile [hunk 5]
```diff
@@ -72,6 +75,20 @@ releases:
...
+
+.PHONY: publish-stable
+publish-stable: push-to-charmstore
+	$(CHARMCMD) publish $(fullurl)
+
+.PHONY: push-to-charmstore
+push-to-charmstore: clean
+	$(CHARMCMD) login
+	@echo $(CHARMCMD) push . $(CHARMURL)
+	$(eval fullurl = $(shell $(CHARMCMD) push . $(CHARMURL) | grep 'url: ' | awk -F 'url: ' '{print $$2}'))
```
Comment (bac) [inline Makefile]: This step is idempotent, correct? An unchanged charm will return the same url as the previous time it was pushed?

File: Makefile (hunk 3 above)
Comment (bac) [inline Makefile]: One reason for including these push/publish targets in the makefile is that it is a multistep process. Doing the clean first is important to ensure no build artifacts get pushed up. It seems like publishing to stable would be a candidate for removing potential errors.

**Comment by bac** [comment]
:+1:

File: Makefile (hunk 3 above)
Comment (frankban) [inline Makefile]: The documentation describes both manual and automated publishing, so I guess it's ok to have both options.

File: QA.md (hunk 2 above)
Comment (frankban) [inline QA.md]: Done.

File: QA.md (hunk 4 above)
Comment (frankban) [inline QA.md]: I know, do you think that's a problem? We need to mention how to run functional tests, and the fact that we have something internally that we use to run integration tests is not really a secret.

File: Makefile (hunk 5 above)
Comment (frankban) [inline Makefile]: Exactly.

**Comment by urosj** [comment]
LGTM

**Comment by frankban** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-charm-merge

**Pull request merged**